File: 4_non-blocking_p2p_comm/4.2_comm_p2p_nonblocking.f90
program my_prog
   use mpi
   implicit none
   integer :: ierr, nprocs, my_rank
   real a(2), b(2)
   integer req0(2), req1(2)
   
   call mpi_init(ierr)
   call mpi_comm_size(mpi_comm_world, nprocs, ierr)
   call mpi_comm_rank(mpi_comm_world, my_rank, ierr)

   
   IF( my_rank .EQ. 0 ) THEN
      a(1) = 2.0
      a(2) = 4.0
      CALL MPI_ISEND(a, 2, MPI_REAL, 1, 10, MPI_COMM_WORLD, req0(1), ierr) 
      CALL MPI_IRECV(b, 2, MPI_REAL, 1, 11, MPI_COMM_WORLD, req0(2), ierr)
      !b(1) = b(1) + 10.0
      CALL MPI_WAITALL(2, req0, MPI_STATUSES_IGNORE, ierr)
      !b(1) = b(1) + 10.0
   ELSE IF( my_rank .EQ. 1 ) THEN
      a(1) = 3.0
      a(2) = 5.0
      CALL MPI_ISEND(a, 2, MPI_REAL, 0, 11, MPI_COMM_WORLD, req1(1), ierr)
      CALL MPI_IRECV(b, 2, MPI_REAL, 0, 10, MPI_COMM_WORLD, req1(2), ierr)
      CALL MPI_WAITALL(2, req1, MPI_STATUSES_IGNORE, ierr)
   END IF
   
   write(*,*) "Hello I am rank number ", my_rank, " and my a array is ", a(1), " ", a(2)
   write(*,*) "Hello I am rank number ", my_rank, " and my b array is ", b(1), " ", b(2)
   call mpi_finalize(ierr)
end program my_prog

File: 4_non-blocking_p2p_comm/Solutions/4.2_comm_p2p_nonblocking.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

void main(int argc, char * argv[]) {
     int err;
     int nprocs, my_rank;
     err = MPI_Init(&argc, &argv);
     MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
     MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
     MPI_Request req0[2], req1[2];
     MPI_Status status[2];
     double a[2], b[2];
        
     if (my_rank == 0) { 
       a[0] = 2.0;
       a[1] = 4.0;
       MPI_Isend(a, 2, MPI_DOUBLE, 1, 10, MPI_COMM_WORLD, &req0[0]);
       MPI_Irecv(b, 2, MPI_DOUBLE, 1, 11, MPI_COMM_WORLD, &req0[1]);
       //b[0] += 10.0;
       MPI_Waitall(2, req0, status);
       b[0] += 10.0;
     }
     else if (my_rank == 1)  {
       a[0] = 3.0;
       a[1] = 5.0;
       MPI_Isend(a, 2, MPI_DOUBLE, 0, 11, MPI_COMM_WORLD, &req1[0]);
       MPI_Irecv(b, 2, MPI_DOUBLE, 0, 10, MPI_COMM_WORLD, &req1[1]);
       MPI_Waitall(2, req1, status);
     }

     printf("My a array is %f, %f and I am proc number %d \n", a[0], a[1], my_rank);
     printf("My b array is %f, %f and I am proc number %d \n", b[0], b[1], my_rank);

     err = MPI_Finalize();
}

